// supertest/index.js
const http = require('http');

function request(app) {
  return {
    post(path) {
      return {
        send(body) {
          return new Promise((resolve, reject) => {
            const server = http.createServer(app);
            server.listen(0, () => {
              const addr = server.address();
              const opts = {
                hostname: '127.0.0.1',
                port: addr.port,
                path,
                method: 'POST',
                headers: { 'Content-Type': 'application/json' }
              };
              const req = http.request(opts, res => {
                let data = '';
                res.on('data', chunk => { data += chunk; });
                res.on('end', () => {
                  server.close();
                  let parsed;
                  try { parsed = JSON.parse(data); } catch { parsed = {}; }
                  resolve({ status: res.statusCode, headers: res.headers, text: data, body: parsed });
                });
              });
              req.on('error', err => { server.close(); reject(err); });
              req.write(JSON.stringify(body));
              req.end();
            });
          });
        }
      };
    }
  };
}
module.exports = request;
